### ISSUE TITLE
Implementation of an example

### ISSUE THREAD

**Issue by karfai** [Xalgorithms/xa-arch#2]
(Refer to https://github.com/Xalgorithms/xa-arch/blob/master/docs/xalgo-examples/bamako-to-montreal/Xalgorithms_Tax%2BDuty_RulesExamples_28may2016XLS.xls)

The linked file contains a set of tables that should be represented in xalgo.

**Comment by karfai** [comment]
(commentary copied from email)

This is a great example! It's helped me understand more about the language. The basic implementation is at https://github.com/Xalgorithms/xa-arch/tree/master/docs/xalgo-examples/bamako-to-montreal. I''ll attempt to explain this implementation in the rest of this response.

First, and foremost, I've added new mutation (INCLUSION and REDUCE) to the language to support this rule. That's precisely what I'd hoped would happen with a more concrete example. I'll update the language spec ASAP to include these changes.

I've made some changes to the data that Joseph provided. This is for simplification as well as implementability.

Changes and assumptions:

1. The attached spreadsheet has a table called VAT. This table has rows which specify BOTH specific countries and rows which are based on inclusion in the set of EU countries. To disambiguate this information, I've broken this table into EU internal VAT and EU external VAT. This allows joining by direct country matching OR inclusion in the EU set.

2. The invoice table includes hs_code. I've removed this from the program in order to show that our Registry could supply the mapping that is described in the table that Joseph supplied in the previous message.

3. I've eliminated the descriptive text columns. They are immaterial to the example.

4. I've assumed that CSL and ST tables are Mali specific, based on the data that appears in them. It's just a naming convention, so it's really not important if I've got this wrong.

Futher: I'll need to type up a line-by-line transcript of the tables on the stack as they would appear when this example were to be run. I'll include that in the example's directory once I've completed it.

**Comment by karfai** [comment]
Response:
>
1. The attached spreadsheet has a table called VAT. This table has rows which specify BOTH specific countries and rows which are based on inclusion in the set of EU countries. To disambiguate this information, I've broken this table into EU internal VAT and EU external VAT. This allows joining by direct country matching OR inclusion in the EU set.
>

[Joseph1]:  I showed WAEMU & !WAEMU -- would you not show EU & !EU?
 
2. The invoice table includes hs_code. I've removed this from the program in order to show that our Registry could supply the mapping that is described in the table that Joseph supplied in the previous message.

3. I've eliminated the descriptive text columns. They are immaterial to the example.

[Joseph1]: But the text columns will certainly be needed for display in the UI. 

4. I've assumed that CSL and ST tables are Mali specific, based on the data that appears in them.

[Joseph1]: Correct (AFAIK), and so they are useful for the example.
 
It's just a naming convention, so it's really not important if I've got this wrong.

[Joseph1]: Not sure what you mean. 

Futher: I'll need to type up a line-by-line transcript of the tables on the stack as they would appear when this example were to be run. I'll include that in the example's directory once I've completed it.

[Joseph1]: Looking at your "table.*" files, I am not sure precisely what part any "rule owner" should "write & publish to the IoR".  
 
Let me know how this example works for you.

[Joseph1]:  Yes and also, not-quite-yet. Maybe I missed seeing in your files the correct presentation in ASCII text of the tables I supplied in the spreadsheet. I am not sure, for example, how to show the VAT table in the form of
foo = [
 { a: 1, b: 2, c: 3 },
 { a: 2, b: 4, c: 6 },
 { a: 3, b: 6, c: 9 },
]
Also, your "example.trace" file seems to make use of all the transactions, but my intention for the time being was that only one thing was being purchased at a time. I included several types of items in my tables only because a full expression of one of these fiscal rules would need to list all the applicable goods/services. But that does not imply a transaction involving multiple goods/services in any given example.

**Comment by karfai** [comment]
Response to simple questions. I expect these need no further discussion. I'll make a second comment for things that will likely lead to more discussion.

> [Joseph1]: I showed WAEMU & !WAEMU -- would you not show EU & !EU?

Yes, for completeness, I would. I didn't in this example because the source spreadsheet didn't include that information. My goal from the "port the rules as they appear".

> [Joseph1]: But the text columns will certainly be needed for display in the UI.

Yes, of course. Just omitted for brevity.

> [Joseph1]: Not sure what you mean.

I mean that I *guessed* on the original of those rules. Had my *guess* been incorrect, I could easily change the naming.

**Comment by karfai** [comment]
> [Joseph1]: Looking at your "table.*" files, I am not sure precisely what part any "rule owner" should "write & publish to the IoR".
>
[Joseph1]: Yes and also, not-quite-yet. Maybe I missed seeing in your files the correct presentation in ASCII text of the tables I supplied in the spreadsheet. I am not sure, for example, how to show the VAT table in the form of
foo = [
{ a: 1, b: 2, c: 3 },
{ a: 2, b: 4, c: 6 },
{ a: 3, b: 6, c: 9 },
]
>

So, in the language's documentation examples, I'm *cheating*. I'm using a Javascript notation to show a *reference* in *xalgo*. In reality, the owners would create these tables and expose them from their repository in the JSON format you see in the table.* files. **This is just the exchange format for the Repository**, they are free to store them internally in their Repository as *whatever they want*. In the reference Repository, we'll do this by keeping JSON files on disk. The *associated reference* will be done as you see in this example (registry:name:version) will be associated with a reference.

Now that I think it through, this example has shown that it is likely going to change to registry:scope:name:version (xalgo:mali_taxation:mail_cst:1234) since the *Registry* will likely be our central one.

>Also, your "example.trace" file seems to make use of all the transactions, but my intention for the time being was that only one thing was being purchased at a time. I included several types of items in my tables only because a full expression of one of these fiscal rules would need to list all the applicable goods/services. But that does not imply a transaction involving multiple goods/services in any given example.

Yeah... I realized this after I got halfway through the example. I kept up thinking that they were a single transaction so that in the example I could show heterogeneity.
